### PR TITLE
Allow Enums in DataMaps

### LINKED ISSUE #504: Enum-Array/Collection fails to serialize to XML String
The implementation in #475 has a bug in that way, that Arrays/Collections of Enums fail to serialize, as the ```DataMap``` explicitly excluded Enums (```Preconditions.checkArgument(!classInfo.isEnum());```)

### PR DESCRIPTION
**Pull request by geri-m**
Fixes #504. 

- Fix issue in ```DataMap``` that caused the problem
- A bunch of tests was added; not only for Enum-Array issue, but other data structures as well
- Replace deprecated ```junit.framework``` and use annotations

### CODE REVIEW

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 1]
```diff
@@ -44,8 +75,205 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+   * We need to method to read the file into a string
+   *
+   * @param file file name in the resource folder that will be parsed
+   * @return content of the File as String.
+   * @throws IOException in case the file was not able to be parsed
+   * @see <a href="https://goo.gl/CJ4v7Z">Stackoverflow</a>
+   */
+  private String readFile(String file) throws IOException {
+    ClassLoader classLoader = getClass().getClassLoader();
+    BufferedReader reader = new BufferedReader(new FileReader(classLoader.getResource(file)
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: This needs an explicit encoding. Don't rely on the platform default encoding.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 2]
```diff
@@ -44,8 +76,209 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+        .getFile()));
+    String line;
+    StringBuilder stringBuilder = new StringBuilder();
+
+    try {
```
[the remark this answers is not in the document]
Comment (elharo) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: Looks good

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 3]
```diff
@@ -44,8 +75,205 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+    while (atomParser.parseNextEntry() != null) {
+      counter++;
+    }
+
+    atomParser.close();
+    assertEquals(62, counter);
+  }
+
+  /**
+   * We need to method to read the file into a string
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: Don't use first person in javadoc. Consider, "read the file into a string". But you should probably just use Guava and

URL url = Resources.getResource(file);
String text = Resources.toString(url, Charsets.UTF_8);

File: google-http-client-xml/src/test/resources/heise-atom.xml [hunk 4]
```diff
@@ -0,0 +1 @@
...4189356</id><updated>2018-10-11T18:32:00+02:00</updated><published>2018-10-11T18:32:00+02:00</published><link rel="alternate" type="text/html" href="https://www.heise.de/newsticker/meldung/Amazon-KI-zur-Bewerbungspruefung-benachteiligte-Frauen-4189356.html?wt_mc=rss.ho.beitrag.atom" /><summary type="html">Eigentlich wollte Amazon eine Software entwickeln, die unter Bewerbern automatisch die besten findet. Der Algorithmus hatte aber unerwünschte Nebenwirkungen.</summary><content type="xhtml"><div xmlns="http://www.w3.org/1999/xhtml"><a href="https://www.heise.de/newsticker/meldung/Amazon-KI-zur-Bewerbungspruefung-benachteiligte-Frauen-4189356.html?wt_mc=rss.ho.beitrag.atom" title="Amazon: KI zur Bewerbungsprüfung benachteiligte Frauen"><img src="https://heise.cloudimg.io/width/264/q80.png-lossy-80.webp-lossy-80.foil1/_www-heise-de_/imgs/18/2/5/1/8/8/6/1/shutterstock_763283053-09be1eb5707c5ea5.jpeg" alt="Amazon: KI zur Bewerbungsprüfung benachteiligte Frauen" /></a><p>Eigentlich wollte Amazon eine Software entwickeln, die unter Bewerbern automatisch die besten findet. Der Algorithmus hatte aber unerwünschte Nebenwirkungen.</p></div></content></entry><entry><title type="text">Schlappe
```
Comment (elharo) [inline google-http-client-xml/src/test/resources/heise-atom.xml]: There may be copyright issues that prevent us from checking this in.

File: google-http-client/src/main/java/com/google/api/client/util/DataMap.java [hunk 5]
```diff
@@ -43,7 +43,8 @@
   DataMap(Object object, boolean ignoreCase) {
     this.object = object;
     classInfo = ClassInfo.of(object.getClass(), ignoreCase);
-    Preconditions.checkArgument(!classInfo.isEnum());
+    // Removed due to issue #504
```
Comment (elharo) [inline google-http-client/src/main/java/com/google/api/client/util/DataMap.java]: don't comment out; delete it or fix it.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java (hunk 3 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: Definitely more elegant than the current implementation, will use guava.

File: google-http-client-xml/src/test/resources/heise-atom.xml (hunk 4 above)
Comment (geri-m) [reply google-http-client-xml/src/test/resources/heise-atom.xml]: Replaced with custom feed and added more validations

File: google-http-client/src/main/java/com/google/api/client/util/DataMap.java (hunk 5 above)
Comment (geri-m) [reply google-http-client/src/main/java/com/google/api/client/util/DataMap.java]: Removed commend out code

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java (hunk 1 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: Switched to guava implementation

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 6]
```diff
@@ -114,19 +133,57 @@ public void testParse_enumAttributeType() throws Exception {
...
       Xml.parseElement(parser, xml, namespaceDictionary, null);
       // fail test, if there is no exception
       fail();
-    } catch (final IllegalArgumentException e){
+    } catch (final IllegalArgumentException e) {
       assertEquals("given enum name ENUM_3 not part of enumeration", e.getMessage());
     }
+  }
+
+  public enum AnyEnum {
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: Changed to private where possible

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 7]
```diff
@@ -1,64 +1,54 @@
...
+      "://www.w3.org/2005/Atom\"><elementEnum>ENUM_3</elementEnum></any>";
+  private static final String XML_ENUM_ELEMENT_ONLY_NESTED = "<?xml version=\"1.0\"?><any " +
+      "xmlns=\"http://www.w3.org/2005/Atom\"><elementEnum>ENUM_2<nested>something</nested" +
+      "></elementEnum></any>";
 
   @SuppressWarnings("cast")
-  public void testParse_anyType() throws Exception {
+  @Test
+  public void testParseAnyType() throws Exception {
     AnyType xml = new AnyType();
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: this variable looks unused

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 8]
```diff
@@ -82,10 +72,34 @@ public void testParse_anyType() throws Exception {
...
+  @Test
+  public void testParseToEnumElementTypeWithNestedElement() throws Exception {
+    assertEquals(XML_ENUM_ELEMENT_ONLY, testStandardXml(XML_ENUM_ELEMENT_ONLY_NESTED));
+  }
+
+  /**
+   * Private Method to handle standard parsing and mapping to {@link AnyTypeEnumElementOnly}
+   *
+   * @param xmlString XML String that needs to be mapped to {@link AnyTypeEnumElementOnly}
+   * @return Returns the serialized string of the XML Objects
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: returns

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 9]
```diff
@@ -82,10 +72,34 @@ public void testParse_anyType() throws Exception {
...
+   * @throws Exception Thrown if there as an issuing when processing the XML.
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: thrown if there is an issue processing the XML
no period

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 10]
```diff
@@ -95,9 +109,14 @@ public void testParse_enumElementType() throws Exception {
     ByteArrayOutputStream out = new ByteArrayOutputStream();
     serializer.setOutput(out, "UTF-8");
     namespaceDictionary.serialize(serializer, "any", xml);
-    assertEquals(XML_ENUM_ELEMENT_ONLY, out.toString());
+    return out.toString();
+
   }
 
+  /**
+   * The purpose of this test is to parse an XML attribute to an objects's member variable
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: member variable --> field

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 11]
```diff
@@ -114,19 +133,60 @@ public void testParse_enumAttributeType() throws Exception {
     assertEquals(XML_ENUM_ATTRIBUTE_ONLY, out.toString());
   }
 
+  /**
+   * The purpose of this test is to parse an XML element to an objects's member variable, whereas
```
Comment (elharo) [inline google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: ditto

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 12]
```diff
@@ -44,8 +79,233 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+
+    // validate feed 6
+    entry = (FeedEntry) atomParser.parseNextEntry();
+    assertNull(entry);
+
+    atomParser.close();
+  }
+
+  /**
+   * Feed Element to map the XML to
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 13]
```diff
@@ -44,8 +79,233 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+    @Key
+    private Author author;
+    @Key
+    private String id;
+    @Key
+    private FeedEntry[] entry;
+  }
+
+  /**
+   * Author Element as part of the {@link Feed} Element to map the XML to (as this is sub-element,
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 14]
```diff
@@ -44,8 +79,233 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+   * this needs to be public)
+   */
+  public static class Author {
+    @Key
+    private String name;
+  }
+
+  /**
+   * Link Element as part of the {@link Feed} Element to map the XML to (as this is sub-element,
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java [hunk 15]
```diff
@@ -44,8 +79,233 @@ public void subtestSetSlugHeader(HttpHeaders headers, String expectedValue, Stri
...
+   * this needs to be public)
+   */
+  public static class Link {
+    @Key("@href")
+    private String href;
+  }
+
+  /**
+   * Entry Element to cover the Entries of a Atom {@link Feed}  (as this is sub-element,
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/AtomTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java [hunk 16]
```diff
@@ -82,10 +72,34 @@ public void testParse_anyType() throws Exception {
     assertEquals(XML, out.toString());
   }
 
-  public void testParse_enumElementType() throws Exception {
-    XmlEnumTest.AnyTypeEnumElementOnly xml = new XmlEnumTest.AnyTypeEnumElementOnly();
+  /**
+   * The purpose of this test is to parse an XML element to an objects's member variable
```
[the remark this answers is not in the document]
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 8 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 7 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: No exactly sure, which variable you refer to. The ```xml``` is used during validation. Could you pls clarified this?

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 9 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 10 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 11 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed.

File: google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java (hunk 7 above)
Comment (geri-m) [reply google-http-client-xml/src/test/java/com/google/api/client/xml/XmlEnumTest.java]: fixed, by removing irrelvant annotation.

**Pull request merged**